### ISSUE TITLE
Amazon Linux AMI Platform Branches Retire on June 30, 2022

### ISSUE THREAD

**Issue by jakubgs** [dap-ps/infra-dapps#22]
There's still a bunch of time before this is relevant, but the Amazon Linux images we use will be retired on 2022/06/30:

```
In order to ensure that our customer’s Elastic Beanstalk environments are aligned with the most current support offered by Amazon Web Services, Elastic Beanstalk is announcing today that platform branches [1] based on Amazon Linux AMI (AL1) will retire on June 30, 2022.

* What does retire mean, and what will happen after June 30, 2022?

Starting on June 30, 2022, retired platform branches will no longer be available for new environments on Elastic Beanstalk. Customers can continue to operate existing environments running on retired platform branches but these branches will no longer be managed and supported by Elastic Beanstalk. For more information see Elastic Beanstalk platform support policy [2] in the AWS Elastic Beanstalk Developer Guide.

* What do I need to do now?

You have up to 12 months to upgrade your Elastic Beanstalk environments configured on platform branches running on the Amazon Linux AMI (AL1) operating system. By updating your environments to use the latest supported platform branch version on Amazon Linux 2, your applications will benefit from long term support, which includes important security, performance, and functionality enhancements. For more information, see Migrating your Elastic Beanstalk Linux application to Amazon Linux 2 [3] in the AWS Elastic Beanstalk Developer Guide

* How does this announcement relate to prior retirement announcements?

We’re consolidating all of the previously announced retirement schedules for Elastic Beanstalk platform branches based on Amazon Linux AMI (AL1). So all of the Amazon Linux AMI (AL1)-based platforms branches will retire on the same date, June 30, 2022.

* Which of my Beanstalk environments are impacted?

To view which of your Elastic Beanstalk environments are impacted in the US-EAST-1 Region:

1. Open the Elastic Beanstalk console [4], and in the Regions list, select your AWS Region.
2. In the navigation pane, choose Environment.
3. The Environments page lists all of your Elastic Environments in the selected Region.
4. If there are any environments in the selected region that are configured with a platform branch that is retired or scheduled for retirement, a message will display at the top of the console page to notify you.
5. The Environments page also includes a Platform state column. The current platform state of each environment is listed beneath this column.

You can also use the "AWS CLI describe-environments" command to display a filtered list of all environments that are configured with Amazon Linux AMI (AL1). See the following example command.

aws elasticbeanstalk describe-environments --region us-east-1 --query 'Environments[?PlatformArn.contains(@,64bit Amazon Linux/)].EnvironmentName' —output text
```

This isn't high priority and won't be an issue for a while, but I will have to get to this eventually. Added to calendar.

**Comment by jakubgs** [comment]
Some more info:

* https://docs.aws.amazon.com/elasticbeanstalk/latest/dg/using-features.migration-al.html
* https://docs.aws.amazon.com/elasticbeanstalk/latest/dg/using-features.platform.upgrade.html

**Comment by jakubgs** [comment]
I looked at the options and saw no way to do it in the UI:

![image](https://user-images.githubusercontent.com/2212681/158471077-16e65a7c-83ac-4227-a077-2ace33fdda8c.png)

If you click "Change" it's already on the "Recommended" version of AMI:

![image](https://user-images.githubusercontent.com/2212681/158471186-5923328c-fb2f-4aca-b157-f036daa11746.png)

I think the issue is that the deployment needs to be built with Amazon Linux 2 as a target.

**Comment by jakubgs** [comment]
According to the docs:

>[Migrating a Linux application to Amazon Linux 2](https://docs.aws.amazon.com/elasticbeanstalk/latest/dg/using-features.migration-al.html) requires a blue/green deployment, because Amazon Linux 2 platform versions are incompatible with previous Amazon Linux AMI platform versions.

https://docs.aws.amazon.com/elasticbeanstalk/latest/dg/using-features.platform.upgrade.html

So it looks like indeed you cannot upgrade the app in place, and you have to create a new environment with the new version.

**Comment by jakubgs** [comment]
I think the proper method to do this upgrade would go something like this:

1. Save the current environment by renaming it in Terraform:
    ```sh
    terraform state mv module.dev_env module.dev_env_backup
    ```
2. Upgrade the `stack_name` for the environment to use Amazon Linux 2:
    https://github.com/dap-ps/infra-dapps/blob/da383e52a17db1b6f5e21f162649264e345c4ee6/variables.tf#L35-L39
3. Create a new environment using Terraform:
    ```sh
    terraform apply --target=module.dev_env
    ```
4. Once new environment works destroy the backup of the old env:
    ```sh
    terraform destroy --target=module.dev_env_backup
    ```

Of course before starting this you'll have to build the current version of [discover](https://github.com/dap-ps/discover) React app ZIP and deploy it to the `dev` env.

Once we figure out this process on `dev` we can repeat it more smoothly with `prod`. I expect there to be issues with duplicate DNS entries and possibly certificates, but not sure.

**Comment by jakubgs** [comment]
One issue that you might encounter is that you don't have SSH access to the hosts as there can be only one key specified:
https://github.com/dap-ps/infra-dapps/blob/f324f6c926c881505543ea8db12ad8978d964416/prod.tf#L68
https://github.com/dap-ps/infra-dapps/blob/f324f6c926c881505543ea8db12ad8978d964416/modules/aws-eb-env/main.tf#L50
https://github.com/dap-ps/infra-dapps/blob/f324f6c926c881505543ea8db12ad8978d964416/main.tf#L43-L46
But I guess it's fine if you temporarily replace that key if you need SSH access.

**Comment by jakubgs** [comment]
Also, I bootstrapped the DB hosts for you so they include your SSH key:
```sh
ANSIBLE_PYTHON_INTERPRETER=/usr/bin/python3 ap ansible/bootstrap.yml -t role::bootstrap:users
```
Needed to change interpreter because the MongoDB Ansible role works only with `python2` and that's set in `ansible.cfg`.

Also had to fix some other issues:

- [f324f6c9](https://github.com/dap-ps/infra-dapps/commit/f324f6c9) - bump Terraform to 1.1.0
- [f78ca158](https://github.com/dap-ps/infra-dapps/commit/f78ca158) - fix host bootstrapping by fixing role name
- [d545adce](https://github.com/dap-ps/infra-dapps/commit/d545adce) - inventory: update IP of db.ev host

**Comment by jakubgs** [comment]
We just had an outage of https://dap.ps/ due to some AWS issues:



Time | Type | Details
-- | -- | --
2022-08-19 23:36:32 UTC+0200 | WARN | Environment health has transitioned from Severe to Degraded. 100.0 % of the requests to the ELB are failing with HTTP 5xx (8 minutes ago). ELB health is failing or not available for 1 out of 2 instances.
2022-08-19 22:47:35 UTC+0200 | WARN | Environment health has transitioned from Degraded to Severe. 100.0 % of the requests to the ELB are failing with HTTP 5xx (6 minutes ago). ELB health is failing or not available for all instances.
2022-08-19 22:46:35 UTC+0200 | WARN | Environment health has transitioned from Severe to Degraded. 100.0 % of the requests to the ELB are failing with HTTP 5xx (9 minutes ago). ELB health is failing or not available for 1 out of 2 instances.
2022-08-18 21:23:57 UTC+0200 | WARN | Environment health has transitioned from Degraded to Severe. 100.0 % of the requests to the ELB are failing with HTTP 5xx (6 minutes ago). ELB health is failing or not available for all instances.

And it looks like it was flapping a lot, essentially since `2022-08-12 10:17:28 UTC+0200`, or at least that far the logs go.

**Comment by jakubgs** [comment]
I found https://dap.ps/ returning 503 errors, and restarting the Elastic Beanstalk environment fixed it:

![image](https://user-images.githubusercontent.com/2212681/185975808-4fe75772-5471-4cff-9dfa-4a1b1d0647e9.png)

At least for now.

**Comment by jakubgs** [comment]
While investigating submissions issues I tried to restart the dev environment which did not give good results:

![image](https://user-images.githubusercontent.com/2212681/195533455-f993405b-a58c-44b9-b088-e15efd9d7f69.png)

I have given up on fixing the `dev-dap-ps-app` environment as it has been stuck at that for ages, so I terminated it.
I've re-created it as Amazon 2 environment and I'm trying to deploy the dev app to it:

![image](https://user-images.githubusercontent.com/2212681/195533607-fab81c74-ef02-4106-ba0f-f6fd6412418a.png)

**Comment by jakubgs** [comment]
I identified an issue with starting the dev app:
```
[ec2-user@ip-10-0-106-11 current]$ sudo systemctl status web.service
● web.service - This is web daemon
   Loaded: loaded (/etc/systemd/system/web.service; enabled; vendor preset: disabled)
   Active: failed (Result: start-limit) since Thu 2022-10-13 08:24:49 UTC; 1h 49min ago
  Process: 4334 ExecStopPost=/bin/sh -c  (code=exited, status=0/SUCCESS)
  Process: 4333 ExecStopPost=/bin/sh -c rm -f /var/pids/web.pid (code=exited, status=0/SUCCESS)
  Process: 4311 ExecStartPost=/bin/sh -c systemctl show -p MainPID web.service | cut -d= -f2 > /var/pids/web.pid (code=exited, status=0/SUCCESS)
  Process: 4310 ExecStart=/bin/sh -c npm start (code=exited, status=1/FAILURE)
 Main PID: 4310 (code=exited, status=1/FAILURE)

Oct 13 08:24:49 ip-10-0-106-11.ec2.internal systemd[1]: Unit web.service entered failed state.
Oct 13 08:24:49 ip-10-0-106-11.ec2.internal systemd[1]: web.service failed.
Oct 13 08:24:49 ip-10-0-106-11.ec2.internal systemd[1]: web.service holdoff time over, scheduling restart.
Oct 13 08:24:49 ip-10-0-106-11.ec2.internal systemd[1]: Stopped This is web daemon.
Oct 13 08:24:49 ip-10-0-106-11.ec2.internal systemd[1]: start request repeated too quickly for web.service
Oct 13 08:24:49 ip-10-0-106-11.ec2.internal systemd[1]: Failed to start This is web daemon.
Oct 13 08:24:49 ip-10-0-106-11.ec2.internal systemd[1]: Unit web.service entered failed state.
Oct 13 08:24:49 ip-10-0-106-11.ec2.internal systemd[1]: web.service failed.
```
If would fail due to some relative path problem with `nodemon`:
```
[ec2-user@ip-10-0-106-11 current]$ npm start

> status-discover-backend@1.0.0 start /var/app/current
> nodemon server.js

internal/modules/cjs/loader.js:905
  throw err;
  ^

Error: Cannot find module '../lib/cli'
Require stack:
- /var/app/current/node_modules/.bin/nodemon
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:902:15)
    at Function.Module._load (internal/modules/cjs/loader.js:746:27)
    at Module.require (internal/modules/cjs/loader.js:974:19)
    at require (internal/modules/cjs/helpers.js:101:18)
    at Object.<anonymous> (/var/app/current/node_modules/.bin/nodemon:3:13)
    at Module._compile (internal/modules/cjs/loader.js:1085:14)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
    at Module.load (internal/modules/cjs/loader.js:950:32)
    at Function.Module._load (internal/modules/cjs/loader.js:790:12)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:75:12) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [ '/var/app/current/node_modules/.bin/nodemon' ]
}
```
I've fixed it by running `sever.js` without `nodemon`: https://github.com/dap-ps/discover/commit/f7688de8

**Comment by jakubgs** [comment]
I managed to get it to load:

![image](https://user-images.githubusercontent.com/2212681/195587436-6ecaa812-129b-4276-8b7f-ee2b025b2a4c.png)

But the IPFS version request is failing:

![image](https://user-images.githubusercontent.com/2212681/195587545-4ea94e5c-2cdd-4d8f-b7f3-1850ec1ea95c.png)

Seems like Infura IPFS endpoint now requires auth:
```
curl "https://ipfs.infura.io:5001/api/v0/version?number=<value>&commit=<value>&repo=<value>&all=<value>" \
    -X POST \
    -u "PROJECT_ID:PROJECT_SECRET"
```
https://docs.infura.io/infura/networks/ipfs/http-api-methods/version

**Comment by jakubgs** [comment]
It appears the support for unauthorized IPFS requests was removed on 10th of August 2022:
https://status.infura.io/incidents/qkv2jyw0h1vp

![image](https://user-images.githubusercontent.com/2212681/195588189-1257cf81-c69d-48fe-b219-29ecccddbe52.png)

https://blog.infura.io/post/ipfs-public-api-and-gateway-deprecation

**Comment by jakubgs** [comment]
One option would be to use our public https://ipfs.status.im/ endpoint, but it allows API calls only for certain AWS IPs:
https://github.com/status-im/infra-ipfs/blob/master/ansible/roles/ipfs-proxy/templates/ipblock.conf.j2

But we could enable the `/api/v0/version` endpoint publicly since that seems to cause no harm:
```
 > curl -s -XPOST https://ipfs.status.im/api/v0/version 
{"Version":"0.15.0","Commit":"3ae52a4","Repo":"12","System":"amd64/linux","Golang":"go1.18.5"}
```

**Comment by jakubgs** [comment]
I bit the bullet and upgraded the `prod-dap-ps-app` environment as well:

![image](https://user-images.githubusercontent.com/2212681/195813779-622d4b48-a3e1-41de-a7c9-b8e5c7beb53b.png)

And it appears to work after some fiddling with the previously used app ZIP I've downloaded.

Changes I made:
- Included `node_modules` I assembed for the devel environment build.
- Updated JS bundle to use `ipfs.status.im` instead of `ipfs.infura.io`.
- Changed `start` script in `package.json` to be `node server.js` instead of `nodemon server.js`.

But the issue with submission popup not showing up still persists.

**Comment by jakubgs** [comment]
Here are all the `infra-dapps` changes:

- https://github.com/dap-ps/infra-dapps/commit/fab9947a - drop SSH public key for Artur
- https://github.com/dap-ps/infra-dapps/commit/945a3763 - aws-eb-env: upgrade subnet and EB submodules
- https://github.com/dap-ps/infra-dapps/commit/d0b97a11 - aws-eb-env: add correct healthcheck_url
- https://github.com/dap-ps/infra-dapps/commit/e2b1b624 - aws-ec2-instance: upgrade AMI to Ubuntu 22.04
- https://github.com/dap-ps/infra-dapps/commit/6e626b9d - versions.tf: upgrade Terraform version to 1.3.0
- https://github.com/dap-ps/infra-dapps/commit/994cb970 - variables.tf: drop unused image_name var
- https://github.com/dap-ps/infra-dapps/commit/b0cb0ae2 - variables.tf: upgrade to Amazon Linux 2 EB image

And here are the `discover` app changes:

- https://github.com/dap-ps/discover/commit/bb1af594 - add shell.nix file to define build environment
- https://github.com/dap-ps/discover/commit/209ad6d2 - run yarn.install in back-end subfolder as well
- https://github.com/dap-ps/discover/commit/86be9230 - makefile: fix node_modules install for back-end
- https://github.com/dap-ps/discover/commit/f7688de8 - start backend without using nodemon
- https://github.com/dap-ps/discover/commit/c959f7f4 - use ipfs.status.im gateway instead of Infura one

Another BIG problem that I see no simple solution for is that Ropsten is discontinued, and that is used in the dev env.
We can't test changes in the `dev-dap-ps-app` environment because Ropsten is broken and Goerli has no contract set up.

**Comment by jakubgs** [comment]
I consider this done. I'll open a separate issue for the submission issues.
